### ISSUE TITLE
Remove tool filtering

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#8464]
##### Issue type
 - Removal request

<https://github.com/minetest/minetest/pull/8458/files> reveals a weird obscure feature in Minetest: Tool filtering.

>                    -- Only drop if using a tool whose name contains
>                    -- "default:shovel_".
>                    tools = {"~default:shovel_"},

So, it is clear that this feature just looks at the itemstring of said items.

I think this feature is completely misguided and is wrong on so many levels. Mods cannot defend against this, nor can they reasonably react. For example, there is no guarantee that an item is an axe only because the string “axe” is inside it. It might be something you just did not expect at all.
 Just looking at the itemstring will potentially lead to many other unexpected results. There is no naming convention for itemstrings that goes beyond the “modname:itemname” form.
This feature seems to date back to 2012 and was left undocumented since then. I have never seen it being used in the wild, and I want it to die for good. I am not sure, maybe this feature is even older than groups, that would explain why it exists.

Groups are clearly much better suited to implement the example use case of “drop by shovels only”.

I therefore propose to completely remove this itemstring-based tool filtering.

**Comment by paramat** [comment]
We're free to remove it as it has not ever been documented in a stable release.
My initial thought is to remove it and not necessarily replace it, consider that later, as no-one has suggested such a feature and i don't see it as essential.

**Comment by Wuzzy2** [comment]
Yes, just remove it without replacement, it's just too obscure.

**Comment by paramat** [comment]
:+1:

**Comment by SmallJoker** [comment]
It existed for years and now you suddenly have a problem with it?
To add more variety to the game it's surely helpful to define the drops to a specific pattern of the mod's registered tool names.
:-1:

**Comment by ClobberXD** [comment]
The technique of tool-filtering by comparing itemstrings is highly prone to problems as explained in the first post of this issue. While I'm all for variety in gameplay, I'd at least prefer a better alternative to string comparison - Wuzzy has already suggested an alternative in their first post, which is making use of groups. I think groups can be put to good use for this feature too.

An example where the current feature would fall short: a mod wants to make gravel drop diamonds when dug by a pick(axe), and adds `tools = {"pick"}`. Imagine another mod registers an item `mod:color_picker`. If this tool could somehow dig gravel, the resulting drop would include diamonds, because the tool used to dig the node contains `pick`. This is a stupid example, but is not as improbable as we might think.

**Comment by paramat** [comment]
I'd be fine with a new method that uses groups, string matching seems a poor implementation.
It existed for years but almost no-one was aware of it.

**Comment by Wuzzy2** [comment]
> It existed for years and now you suddenly have a problem with it?

Dude! It was an obscure, undocumented feature and obviously I didn't even know it existed at all. After it got documented, I immediately spoke out against it.
And I have never seen this tool filtering being used by anyone. Which is not surprising since it was undocumented.

Can we please agree to first throw away this broken design for now and *after that* discuss about possible better implementations?

It is pointless to deliberately keep a badly designed API that nobody uses.

**Comment by paramat** [comment]
It might be a good idea to remove it quickly now before anyone gets used to it, instead of waiting months/years until a better implementation is added.

**Comment by Wuzzy2** [comment]
Mark it “high priority” then.

**Comment by paramat** [comment]
Not really.
Now waiting for more core dev opinins for what to do.

**Comment by ashtrayoz** [comment]
It is wrong to say that this feature is not in use.  As noted in #7231, I have dozens of nodes that depend on this.  That is why I started documenting it ("SmallJoker added the **Documentation** label on 10 Apr 2018"). 

But I agree that the feature can be improved, which is why I was seeking feedback from other users ("SmallJoker added the **User feedback needed** label on 16 Apr 2018").  My pull request tried to fit in with what already existed, rather than being too radical.  It also offered either backwards compatibility or a (slightly) improved API - it was my opinion that both would probably not be easy to achieve.  Without user feedback, the PR languished half-way between a feature-pull-request and a documentation-pull-request.  Thanks go to @Paramat for picking it up again.

The suggestion that "mod:colour_picker" would be misrecognised as a "pick" is overcome in the same way as any #other regexp-based programming problem - include some form of anchor in the string.  The most obvious case is ":pick_", which is why most tools are named "pick_mese" or "pick_stone", and not "stone_pick".  This is not perfect, but allows a mod to "defend" against the most obvious problems.  If the mod name is included before the colon, then the defense is as complete as it can be.

But that does not mean that we have to keep item-string based filtering.  If we are looking to break compatibility, then I have to ask (as I did in 7231) "What do other people think?", or perhaps "What do other modders want?"

Personally, I like the idea from @Wuzzy2 of using groups, even if it has its own problems.  For example, there are lots of crafting recipes which try to achieve "you can craft this out of leaves, except palm leaves" using the group mechanism, and it is a mess.  Using groups essentially changes the semantics from "called a shovel" to "is a shovel", which is a step in the right direction.

I still think that "NOT" needs to be part of the filtering mechanism, whether it is group-based or something else.  There are multiple ways that the API and syntax could be constructed, and I would probably be willing to put some time into this, if users are willing to provide feedback about the pros and cons of the issue.

And I have some users who never cut down trees with axes if they want planks - saws are so much more efficient!

**Comment by Wuzzy2** [comment]
> As noted in #7231, I have dozens of nodes that depend on this.

Which nodes?

But even if, that doesn't change the fact that this is a shoddily implemented feature that should have never been added. I honestly don't even care if some mod actually does depend on it. Any mod that depends on this is doomed to be buggy. There is no reason to further support this. As it was never documented in a stable release, I'd argue this is not even a true removal from the official API. So modders are fully responsible when something breaks about an obscure, broken-by-design, undocumented feature. Any function that is not documented does not exist. :P

> Personally, I like the idea from @Wuzzy2 of using groups, even if it has its own problems.

What you talk about is just about a missing feature (mainly a logical NOT), but it is not an inherent design flaw.

**Comment by ClobberXD** [comment]
I'm sure logical operators like AND, NOT, and OR can be implemented with groups as well.

**Comment by paramat** [comment]
Because this is in use, if we remove the current filtering with the intention of replacing it with another system, the new system should be added immediately.
This issue doesn't specify 'immediately replace' so i'm removing my support.
Whatever new system we use i don't support any additional capability, matching to a group or a list of groups is enough to cover current capability.

**Comment by Wuzzy2** [comment]
First of all, at the very least, the current string filtering should be marked as deprecated, if a proper fix doesn't make it into the next release. Modders should be discouraged from using this.

Now, here's some ideas to replace the string-based filtering:

Matching multiple tools: Match items based on group identity, not on string. If you write “group:shovel”, it will match all items of the shovel group. Easy-peasy. It's done in like literally everywhere else in the whole API. Consistency is great!

About the logical NOT: Use a magic character in front of either the full itemstring or the `group:*` string. This means, it matches on all tools EXCEPT the tool with the given itemstring or group identity. For example: “!default:shovel_wood” matches on any item except `default:shovel_wood”. The syntax should be not any more complex than that.
By the way, do we even support the logical NOT right now? Because it doesn't seem mentioned in the documentation … If it turns out the logical NOT is not even supported, then we can ignore it safely. :-)

@paramat: Also, I like to note that @ashtrayoz has still not provided any evidence for the claim this feature is used “in a lot of nodes”.

**Comment by SmallJoker** [comment]
Replace it with Lua-regex.
 - `default:.*_steel` steel stuff
 - `~default:shovel_.*` inverse matching

**Comment by Wuzzy2** [comment]
Sigh.
You missed the point …
I want the string matching / regex bullshit to be removed entirely. Have you not read the first post?

Perhaps I should explain it differently …

Reasons against regex:

* This is the only place in which regexes are used. Everywhere else in the API, we use groups. This inconsistency makes no sense
* The whole assumption the regex matching makes about itemstrings is deeply flawed: It assumes that itemstrings follow a logical pattern and some predictable rules. They don't. Apart from the `modname:itemname` scheme, itemstrings are pretty much a free-for-all. Every mod can use itemstrings however they like. Some mods like to use underscores. Some others don't. Shovels might or might not have the word “shovel” in their itemstring. Steel stuff might or might not have “steel” in their itemstring. My point is, the itemstrings are simply not made for string matching. You can try to apply string matching, but you will just get nonsensical results. This is too unreliable to build a string-matching on.
* Regex matching does not win us anything we can not already do with groups. Therefore, regex matching is just redundant and overcomplicates things. Groups actually carry meaning. Itemstrings don't. Want to match all shovels? Just look for the `shovel` group. Oh, MTG, doesn't have the `shovel` group yet? Just fix MTG, then!
* It's not future-proof. Since this only looks at the naked string, be prepared for all sorts of surprises when the mod you depend on adds new items that somehow match your regex, but you definitely did not intend those to be matched. Let's say you match `example:pick` for all pickaxes (e.g. `example:pickstone`, `example:pickwood`, etc.). Fast-forward to version 2 of `example`, which adds `example:pickle`, which is something completely different, yet still matches … Had we just looked on the group `pickaxe` instead, which all pickaxes have, then it would still work.
* This leads to all sorts of hilarious surprises/bugs. If you match, like `default:shovel_`, you might also match `sillydefault:shovel_butter`.

TL;DR: Regex matching for itemstring is a terrible idea, invites a swarm of bugs, and doesn't add anything we cannot already do with groups. Regex matching can and should be dropped. Its existence is making the Lua API worse, not better.

As for the logical NOT, I would keep it super-simple and just allow to negate *concrete* items or entire groups, without having the full regex baggage applied. I already explained above.

**Comment by rubenwardy** [comment]
I'm against regex matching for this purpose. I suggest deprecating the functionality as a misfeature, and suggest using groups instead

**Comment by Wuzzy2** [comment]
Please make a formal decision on what to do with tool filtering for next release 5.1.0. The longer you wait with dealing this, the worse it will become.

Here's a couple of possible decisions (non-exhaustive):

1. Do absolutely nothing
2. Mark tool filtering as deprecated and strongly discouraged in `lua_api.txt`
3. Remove tool filtering without replacement
4. Remove tool filtering with replacement (e.g. groups-based filtering)

My opinion on the decisions:

1. IMO the absolute worst decision, see reasons above
2. Doesn't really solve the problems, but it's easy to do and at least you can then rightfully blame modders for depending on deprecated features, and modders can no longer blame you :D. It's much better than doing nothing.
3. IMO the best solution that would realistically work. Plus, we have a PR for it.
4. The best solution in theory, but I don't think this is sadly now out of reach since we're in feature freeze. But I don't think the removal of this feature should depend on a replacement anyway. There's just basically 0 mods depending on this, thankfully. If you make a replacement a requirement for removal, that's just an excuse to delay removal/deprecation indefinitely, and the problem will just become worse and worse and worse.

The longer you wait with removing or *at least* deprecating it, the worse it will become. Thankfully, it seems, we still have basically 0 mods depending on this. PLEASE deal with this as long any potential disruption is very low or non-existing.

Thank you for reading this.

**Comment by sfan5** [comment]
Since it's the feature feature freeze, 3 and 4 are out of question.
2 might happen if we feel it's important (probably not), so 1 it is.

> I don't think the removal of this feature should depend on a replacement anyway. [...], that's just an excuse to delay removal/deprecation indefinitely

I sure hope nobody sane agrees with this "We thought this feature sucks and couldn't be bothered, so we removed it entirely. Have fun with your broken code lol." stance.

**Comment by Wuzzy2** [comment]
Note an important argument here is that there is zero evidence of anyone ever using this feature so far. Remove this feature while it's easy. If you're that strict about feature freeze, then please deal with it after release.

To quote Krock again:

> 1. 0 uses in ~270 mods (includes several popular ones)
> 2. Undocumented for a long time (until 5.0.0 or so)
> 3. Drops still work, but different to what you expect if tools were defined.

I totally get your fear of breakage, but this here is special: Your fear is mostly theoretical in this time. I have yet to see ANY evidence of any mod using this feature. I have asked again and again for evidence, but nobody posted anything.

Of course it would be great if we had a replacement ready, but we don't and we're in feature freeze anyway.

What's the reason against deprecation? NOT deprecating it means indirectly endorsing it. Which is BAD. Why?

Because regex-based filtering is completely misguided and promotes bad code and it should have been group-based filtering from the start. And the longer you wait, the higher the risk that modders will actually start adopting it.

You can choose between two evils. The question is, which evil is less bad.

1. Deprecating or removing it. This has the low risk of MAYBE breaking a few mods, but the current evidence for any mods using this does not exist at all
2. Keeping the feature, not deprecating it. This has the risk of modders actually starting to adopt it, making the bad feature more and more enshrined and more impossible to remove. Not even warning modders about the downside of this feature is just making it even worse, when we actually do have core devs that agree it should be at least replaced. Not deprecating it sounds like you actually want to support this feature till the end of time.

Both solutions are ugly. But I think solution 2 is uglier. Solution 1 is only theoretically ugly.

**Comment by celeron55** [comment]
Making sure group based filtering works and then deprecating item name wildcard based filtering seems like the reasonable thing to do.

The name based string matching was added in https://github.com/minetest/minetest/commit/157a4cf18cb9c098f465b8baecd7d2cd5705f2dd#diff-494a8464a5335033fe2460099e805050R158 by kahrl when the drop definition for nodes was changed from a replaceable callback to the drop field that is still in use. Item groups did not exist back then.

**Comment by paramat** [comment]
> Mark tool filtering as deprecated and strongly discouraged in lua_api.txt

I support this, should be done for 5.1.0.

**Comment by paramat** [comment]
Deprecation PR #9009

**Comment by paramat** [comment]
Changing title and adding label to show celeron55's support of adding group based tool filtering.
